Tests/Feature/ConversationEngineFunctionalTest.php

```diff
@@ -0,0 +1,1162 @@
+
+        // $firstMessage = Hi! This is the default welcome message for the Webchat Scenario.
+
+        /**
+         * 2. Given any text utterance (i.e. “hello world”), for an existing user, we should receive
+         * the welcome message response again. To start, the user’s state should be preset to MID_TURN
+         * in the welcome_turn (which is the output state from #1). We should see that both interpreters
+         * are consulted, the new interpreter’s response should be mocked to return an intent collection,
+         * containing a matching “HelloIntent” intent.
+         *
+         * Utterance: Text utterance when types “Hello world”
+         *
+         * Interpreter first intent: OpenDialog interpreter - returns []
+         *
+         * Interpreter second intent: Dialogflow interpreter - returns [HelloIntent]
+         *
+         * Message outcome: Welcome message response
+         */
+
+        $this->app->forgetInstance(ConversationEngineInterface::class);
+        $conversationEngine = resolve(ConversationEngineInterface::class);
+
+        // Set up listeners
+
+        $states = new Map();
+
+        resolve(Dispatcher::class)->listen(IncomingIntentStateUpdate::class, function (IncomingIntentStateUpdate $event) use ($states) {
+            $states->put(get_class($event), $event->getConversationalState());
+        });
+
+        resolve(Dispatcher::class)->listen(OutgoingIntentStateUpdate::class, function (OutgoingIntentStateUpdate $event) use ($states) {
+            $states->put(get_class($event), $event->getConversationalState());
+        });
+
+        resolve(Dispatcher::class)->listen(MessageTemplateStateUpdate::class, function (MessageTemplateStateUpdate $event) use ($states) {
+            $states->put(get_class($event), $event->getConversationalState());
+        });
+
+        // Prepare utterance and state for user
+
+        $utterance = UtteranceGenerator::generateTextUtterance('Hello world!');
+
+        $user = $utterance->getUser();
+
+        $state = (new ConversationalState(Intent::APP, ConversationalState::MID_TURN))
+            ->setScenarioId($scenario->getUid())
+            ->setConversationId(PathSubstitutionHelper::findObject($scenario, '$path:webchat/welcome_conversation')->getUid())
+            ->setSceneId(PathSubstitutionHelper::findObject($scenario, '$path:webchat/welcome_conversation/welcome_scene')->getUid())
+            ->setTurnId(PathSubstitutionHelper::findObject($scenario, '$path:webchat/welcome_conversation/welcome_scene/welcome_turn')->getUid())
+            ->setIntentId(PathSubstitutionHelper::findObject($scenario, '$path:webchat/welcome_conversation/welcome_scene/welcome_turn/_requests/0')->getUid())
+            ->setMessageTemplateId(MessageTemplate::UNDEFINED)
+            ->setConversationSystem(null);
+
+        ConversationalStateReasoner::setConversationalStateForUser($user, $state);
+
+        $user->setRootId('utterance_user');
```

Comment:
It's worth using the constant here instead:

```suggestion
        $user->setRootId(UtteranceAttribute::UTTERANCE_USER);
```

Reply:
fixed